python/cugraph/dask/core.py

```diff
@@ -0,0 +1,184 @@
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+import numba.cuda
+import time
+import os
+from threading import Lock, Thread
+import cugraph
+
+
+class IPCThread(Thread):
+    """
+    This mechanism gets around Numba's restriction of CUDA contexts being
+    thread-local by creating a thread that can select its own device.
+    This allows the user of IPC handles to open them up directly on the
+    same device as the owner (bypassing the need for peer access.)
+    """
+
+    def __init__(self, ipcs, device):
+        """
+        Initializes the thread with the given IPC handles for the
+        given device
+        :param ipcs: list[ipc] list of ipc handles with memory on the
+                     given device
+        :param device: device id to use.
+        """
+
+        Thread.__init__(self)
+
+        self.lock = Lock()
+        self.ipcs = ipcs
+
+        # Use canonical device id
+        self.device = get_device_id(device)
+
+        print("Starting new IPC thread on device %i for ipcs %s" %
+              (self.device, str(list(ipcs))))
+        self.running = False
+
+    def run(self):
+        """
+        Starts the current Thread instance enabling memory from the selected
+        device to be used.
+        """
+
+        select_device(self.device)
+
+        print("Opening: " + str(self.device) + " "
+              + str(numba.cuda.get_current_device()))
+
+        self.lock.acquire()
+
+        try:
+            self.arrs = [ipc.open() for ipc in self.ipcs]
+            self.ptr_info = [x.__cuda_array_interface__ for x in self.arrs]
+
+            self.running = True
+        except Exception as e:
+            logging.error("Error opening ipc_handle on device " +
+                          str(self.device) + ": " + str(e))
+
+        self.lock.release()
+
+        while (self.running):
+            time.sleep(0.0001)
```

Comment:
May adding a comment explaining why you picked 0.0001 will be helpful.

Reply:
I wish this file was in raft so that we address that once. @Iroy30 any thought?

Reply:
To have a while loop continuously checking on self.running status. time 0.0001 is probably an arbitrary small enough time to repeat the condition checking. In cpp, probably a "while(condition);" is enough but python would need to have a function body